packages/datadog-instrumentations/src/mquery.js

```diff
@@ -0,0 +1,88 @@
+'use strict'
+
+const {
+  channel,
+  addHook,
+  AsyncResource
+} = require('./helpers/instrument')
+const shimmer = require('../../datadog-shimmer')
+
+const prepareCh = channel('datadog:mquery:filter:prepare')
+const startCh = channel('datadog:mquery:filter:start')
+const finishCh = channel('datadog:mquery:filter:finish')
+
+const methods = [
+  'find',
+  'findOne',
+  'findOneAndRemove',
+  'findOneAndDelete',
+  'count',
+  'distinct',
+  'where'
+]
+
+const methodsOptionalArgs = ['findOneAndUpdate']
+
+function getFilters (args, methodName) {
+  const [arg0, arg1] = args
+
+  const filters = arg0 && typeof arg0 === 'object' ? [arg0] : []
+
+  if (arg1 && typeof arg1 === 'object' && methodsOptionalArgs.includes(methodName)) {
+    filters.push(arg1)
+  }
+
+  return filters
+}
+
+addHook({
+  name: 'mquery',
+  versions: ['>=5.0.0']
+}, Query => {
+  [...methods, ...methodsOptionalArgs].forEach(methodName => {
+    if (!(methodName in Query.prototype)) return
+
+    shimmer.wrap(Query.prototype, methodName, method => {
+      return function wrappedMqueryMethod () {
+        if (prepareCh.hasSubscribers) {
+          const filters = getFilters(arguments, methodName)
+          if (filters?.length) {
+            prepareCh.publish({ filters })
+          }
+        }
+
+        return method.apply(this, arguments)
+      }
+    })
+  })
+
+  shimmer.wrap(Query.prototype, 'exec', originalExec => {
+    return function wrappedExec () {
+      if (!startCh.hasSubscribers) {
+        return originalExec.apply(this, arguments)
+      }
+
+      const asyncResource = new AsyncResource('bound-anonymous-fn')
```

Comment:
Can you please use `tracingChannel` rather than the outdated `AsyncResource` form?

Reply:
Changed. It simplifies a lot the wrapper but:
  
  a) AFAIK if you are only interested in two events like `start` and `asyncStart` there is no way to avoid creating the rest even specifying a channels object in the `TracingChannel` constructor
  
  b) if you need some other event you cannot publish it directly with the `TracingChannel`. You need to create it separately. Maybe would be handy to allow to publish '`tracing:${nameOrChannels}:${someOtherEvent}`' events like `tracingChannel.publish('someOtherEvent', context)` :)

Reply:
> a) AFAIK if you are only interested in two events like `start` and `asyncStart` there is no way to avoid creating the rest even specifying a channels object in the `TracingChannel` constructor

The others will be no-ops unless there are subscribers, so that's fine.

> b) if you need some other event you cannot publish it directly with the `TracingChannel`. You need to create it separately. Maybe would be handy to allow to publish '`tracing:${nameOrChannels}:${someOtherEvent}`' events calling a method like `tracingChannel.publish('someOtherEvent', context)` :)

Yes, that was an intentional decision to not expand the scope beyond what was strictly necessary for tracing. It's easy to create separate individual channels for any other needs. I would recommend against using the `tracing:` prefix for those additional events though as that's _supposed_ to indicate that the event is one of the core tracing channel event types. The intent is for TracingChannel to _only_ express tracing lifecycle events, but the idea of a helper for publishing to unrelated channels could be revisited.

Reply:
ok, got it!